### ISSUE TITLE
Can `{a: never}` be treated similar to `never`?

### ISSUE THREAD

**Issue by cakoose** [microsoft/TypeScript#42324]
# Suggestion

## 🔍 Search Terms

- "never field"
- "never record"
- "never subsume"

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [?] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [?] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

A value of type `{a: never}` can't be instantiated, right?  What would be the consequences if TS treated `{a: never}` more like `never` and, for example, didn't report an error in function `g2` below ([playground link](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVIEYAKASgC55UQA3EGAbgCgV1s94BzY8+AZwxhZUHeAG9G8SfDgZkMfIm714AehXwAcjnh0YOGIwC+jRqEiwELTLgUAmUhVFQKVWjENMrbfB3s9+gsJiElIycnakymrwAKIwejAUACoAngAOCADkovDOlDR0yoaZ8Fi8lDgYuby8WByoUABGEAgY2hjpWQFCHJkAdEZAA)):

```typescript
declare function f1(): never;
function g1(): string {
    return f1(); // No error
}

declare function f2(): {a: never};
function g2(): string {
    return f2(); // Error: Type '{ a: never; }' is not assignable to type 'string'.
}
```

## 📃 Motivating Example

I have a generic utility type called `Result` for operations that might return a value or might return an error:

```typescript
export type Result<T, E> =
    | {ok: true, value: T}
    | {ok: false, err: E};

// Normal use
const r: Result<number, string> = performOperation1();
if (!r.ok) {
    reportError(r.err);
    return;
}
doSomething(r.values);
```

If I pass in never for the E type parameter, I know that the ok must be true, but TypeScript doesn't:

```typescript
const r: Result<number, number> = performOperation2();
doSomething(r.value);
```

(Note: In this tiny example, `performOperation2` could just return `number` instead of `Result`.  But in my full codebase, all of these "operation" functions conform to an interface, which returns `Result`.)

Someone suggested redefining `Result` with a conditional type ([StackOverflow link](https://stackoverflow.com/a/65697605/163832)):

```typescript
export type Result<T, E> =
    | {ok: true, value: T}
    | ([E] extends [never] ? never : {ok: false, err: E})
```

The downsides:
- I can only do this for unions that I control.
- It adds a bunch of noise to the definition.
- It's tedious and error prone.

What if TS treated `never` in a way such that this simplification happened automatically?  Would that cause soundness/performance/ergonomics issues?

**Comment by RyanCavanaugh** [comment]
Property resolution is *deferred*; if a property of a type is never needed to be computed, it simply isn't. This is important and relevant because a) it's extremely possible for an infinite chain of property paths to exist and b) a `never` can be located at any arbitrary point in those paths. For example

```ts
type IsPassword<T> = T extends "magic" ? never : T;

type Alphabet = "a" | "m" | "i" | "g" | "c";
type PuzzleBox<T extends string> = {
    [K in Alphabet]: PuzzleBox<IsPassword<`${T}${K}`>>
} & { value: T };

declare const p: PuzzleBox<"">;
// a: "agim"
const a = p.a.g.i.m.value;

// b: never
const b = p.m.a.g.i.c.value;
```

There's really no reason to *write* the concrete type `{ a: never }` (it's a long way to write `never`), so it only really makes sense to consider this request in the context of a type computation, but we simply can't exhaustively check all subproperties of an object to see if any of them are `never`.

We also have to account for the possibility that someone wrote `{ a: never }` to indicate that `a` was a *property getter* that throws on access, which is a thing JS has.

**Comment by 00ff0000red** [comment]
I for one had literally used a `never` getter this week for a property that was never meant to be accessed but already existed on the base class. If/when "throw types" are completed and released, I'll definitely update it, but until then, `never` is my only option.

**Comment by cakoose** [comment]
@RyanCavanaugh, thanks for the explanation!

> There's really no reason to _write_ the concrete type `{ a: never }` (it's a long way to write `never`), so it only really makes sense to consider this request in the context of a type computation,

I agree that there's no reason to write the concrete type `{ a: never }`.  In my case, I needed it for a generic type `Result<..., never>`; is that still considered a "type computation"?

I wonder if it's possible to achieve something useful by only simplifying to `never` in simple, predictable cases.  For example:
- unions: remove members that simplify to `never`.
- records: simplify to `never` if any property simplifies to `never`.
- conditional types: simplify to `never` if _both_ branches simplify to `never`.

Would that address this particular problem?  (Ignoring for the moment whether it's "worth it" for TS.)

> We also have to account for the possibility that someone wrote `{ a: never }` to indicate that `a` was a _property getter_ that throws on access, which is a thing JS has.

Oh dang, didn't think about this.  I don't have an ideas for this :-\

In a way, a JS object `{ a:  never }` is sort of like a simple record `{ a: () => never }`, where the function is called implicitly.  And since `() => never` _can_ exist, we can't simplify the whole thing to `never`.

**Comment by RyanCavanaugh** [comment]
> records: simplify to never if any property simplifies to never.

This implies a recursive process that would infinitely descend into a type like the one I showed (or even under some finite-depth algorithm, perform an enormous amount of work and still not produce the right answer)

**Comment by cakoose** [comment]
(This is all kind of moot wrt changing TS, because of the `never` property problem.  I'm only commenting because it's kind of an interesting theoretical problem.  Feel free to ignore.)

Yes, we can't always determine whether a type is equivalent to `never`, especially in a Turing-complete type system.  The best we can do is an approximation that yields either "definitely equivalent to `never`" or "I'm not sure".

And I agree that setting an arbitrary depth limit is a bad way to approximate, since it results in unpredictable behavior for the programmer.  But maybe there's another approximation, similar in spirit to how a Turing-complete language (with recursion, etc) can still have useful type inference?
